src/Hamiltonians/HubbardMomSpace.jl

```diff
@@ -0,0 +1,639 @@
+        step = 2π/S[i]
+        if isodd(S[i])
+            start = -π*(1+1/S[i]) + step
+        else
+            start = -π + step
+        end
+        kr = range(start; step = step, length = S[i])
+        if isodd(S[i])
+            push!(ks_mat,[j for j in kr])
+        else
+            push!(ks_mat,reverse([j for j in kr]))
+        end
+    end
+    kes, ks = dispersion_MomSpace(SVector{D}(ks_mat), geometry, t_mat)
+
+    return HubbardMomSpace{C,D,typeof(address),typeof(geometry),typeof(ks),typeof(kes),typeof(t_mat),
+        typeof(u_mat),typeof(w_mat)}(
+        address, ks, kes, t_mat, u_mat, w_mat, geometry,
+    )
+end
+
+LOStructure(::Type{<:HubbardMomSpace}) = IsHermitian()
+
+function Base.show(io::IO, h::HubbardMomSpace{C}) where {C}
+    io = IOContext(io, :compact => true)
+    println(io, "HubbardMomSpace(")
+    println(io, "  ", starting_address(h), ",")
+    println(io, "  geometry = ", h.geometry, ",")
+    println(io, "  t = ", (h.t), ",")
+    if isnothing(h.u)
+        println(io, "  u = ", zeros(C,C), ",")
+    else
+        println(io, "  u = ", Float64.(h.u), ",")
+    end
+    if isnothing(h.w)
+        println(io, "  w = ", zeros(C,C), ",")
+    else
+        println(io, "  w = ", Float64.(h.w), ",")
+    end
+    print(io, ")")
+end
+
+# Overload equality due to stored potential energy arrays.
+Base.:(==)(H::HubbardMomSpace, G::HubbardMomSpace) = all(map(p -> getproperty(H, p) == getproperty(G, p), propertynames(H)))
+
+starting_address(h::HubbardMomSpace) = h.address
+
+dimension(::HubbardMomSpace, address) = number_conserving_dimension(address)
+
+# offdiaonals =========================================================================== #
+# Holds the offdiagonals for a single-component nearest neighbour one-body term. It's
+# structured like a matric where the first index determines the occupied site in the adress
+# and the second index determines the site the particle will hop to.
+struct HubbardMomSpaceComponentData{C,I1,I2,D,G,A,A1,A2,O1,O2} <: AbstractMatrix{Pair{A,Float64}}
+    geometry::G
+    parent_address::A
+    address1::A1
+    address2::A2
+    u::Union{Float64, Nothing} # interaction strength
+    w::Union{Float64, Nothing} # nearest neighbour interaction strength
+    occmap1::O1
+    occmap2::O2
+
+    function HubbardMomSpaceComponentData{C,I1,I2,D}(
+        geometry::G,
+        parent::A,
+        address1::A1,
+        address2::A2,
+        u::Union{Float64, Nothing},
+        w::Union{Float64, Nothing},
+        occmap1::O1=occupied_mode_map(address1),
+        occmap2::O2=occupied_mode_map(address2),
+    ) where {C,I1,I2,D,G,A,A1,A2,O1,O2}
+        return new{C,I1,I2,D,G,A,A1,A2,O1,O2}(geometry, parent, address1, address2, u, w, occmap1, occmap2)
+    end
+end
+
+function Base.size(data::HubbardMomSpaceComponentData{<:Any,I,I}) where {I}
+    M= num_modes(data.address1)
+    s1, d1 = num_singly_doubly_occupied_sites(data.address1)
+    return  s1 * (s1-1) * (M-2) + d1*(M-1)
```

Comment:
`HubbardMomSpaceComponentData` is `<:AbstractMatrix`. If that's the case, `size` should return two numbers.
If you want it to have a single dimension, it should be `<:AbstractVector` and return a 1-tuple as its size like so:
```suggestion
    return  (s1 * (s1 - 1) * (M - 2) + d1*(M - 1), )
```